core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -52,6 +58,722 @@ static TEE_Result ffa_get_dst(struct thread_smc_args *args,
+	receiver->perm.endpoint_id = endpoint_id;
+	receiver->perm.perm = READ_ONCE(access_perm->perm);
+	receiver->perm.flags = READ_ONCE(access_perm->flags);
+
+	if (!sp_get_session(receiver->perm.endpoint_id)) {
+		free(receiver);
+		return FFA_DENIED;
+	}
+	if (receiver->perm.perm & ~FFA_MEM_ACC_MASK) {
+		free(receiver);
+		return FFA_DENIED;
+	}
+
+	SLIST_INSERT_HEAD(&smem->receivers, receiver, link);
+
+	return FFA_OK;
+}
+
+static void spmc_sp_handle_mem_share(struct thread_smc_args *args,
+				     struct ffa_rxtx *rxtx,
+				     struct sp_session *owner_sp)
+{
+	uint64_t global_handle = 0;
+	int res = FFA_OK;
+	uint32_t ret_w2 = 0;
+	uint32_t ret_w3 = 0;
+
+	cpu_spin_lock(&rxtx->spinlock);
+
+	/*clear top bits*/
+	args->a1 &= 0x0000ffff;
+
+	res = spmc_sp_add_share(rxtx, args->a1, &global_handle, owner_sp);
+	if (!res) {
+		reg_pair_from_64(global_handle, &ret_w3, &ret_w2);
+		args->a3 = ret_w3;
+		args->a2 = ret_w2;
+		args->a0 = FFA_SUCCESS_32;
+	} else {
+		ffa_set_error(args, res);
+	}
+
+	cpu_spin_unlock(&rxtx->spinlock);
+}
+
+static int spmc_sp_add_sp_region(struct sp_mem *smem,
+				 struct ffa_address_range *mem_reg,
+				 struct sp_session *owner_sp,
+				 uint8_t highest_permission)
+{
+	struct sp_ctx *sp_ctx = NULL;
+	uint64_t va = READ_ONCE(mem_reg->address);
+	int res = FFA_OK;
+	uint64_t region_len = READ_ONCE(mem_reg->page_count) * SMALL_PAGE_SIZE;
+	struct mobj *mobj = NULL;
+
+	sp_ctx = to_sp_ctx(owner_sp->ts_sess.ctx);
+
+	/*
+	 * The memory region we try to share might not be linked to just 1
+	 * mobj. Create a new region for each mobj.
+	 */
+	while (region_len) {
+		size_t len = region_len;
+		struct sp_mem_map_region *region = NULL;
+		uint16_t prot = 0;
+		size_t offs = 0;
+
+		/*
+		 * There is already a mobj for each address that is in the SPs
+		 * address range.
+		 */
+		mobj = vm_get_mobj(&sp_ctx->uctx, va, &len,
+				   &prot, &offs);
+
+		if (!mobj)
+			return FFA_DENIED;
+
+		/*
+		 * If we share memory from a SP, check if we are not sharing
+		 * with a higher permission than the memory was originally
+		 * mapped.
+		 */
+		if ((highest_permission & FFA_MEM_ACC_RW) &&
+		    !(prot & TEE_MATTR_UW)) {
+			res = FFA_DENIED;
+			goto out;
+		}
+
+		if ((highest_permission & FFA_MEM_ACC_EXE) &&
+		    !(prot & TEE_MATTR_UX)) {
+			res = FFA_DENIED;
+			goto out;
+		}
+
+		region = malloc(sizeof(*region));
+		region->mobj = mobj;
+		region->page_offset = offs;
+		region->page_count = len / SMALL_PAGE_SIZE;
+
+		if (!sp_has_exclusive_access(region, &sp_ctx->uctx)) {
+			free(region);
+			res = FFA_DENIED;
+			goto out;
+		}
+
+		va += len;
+		region_len -= len;
+		SLIST_INSERT_HEAD(&smem->regions, region, link);
+	}
+
+	return FFA_OK;
+out:
+	mobj_put(mobj);
+	return res;
+}
+
+static int release_mobj(struct sp_mem *smem, struct sp_mem_map_region *region,
+			bool mobj_ffa)
+{
+	uint64_t handle = smem->global_handle;
+	int res = FFA_OK;
+
+	if (mobj_ffa) {
+		if (handle) {
+			mobj_put(region->mobj);
```

Comment:
put the mobj if returning FFA_DENIED later?
maybe move after line 242.

Reply:
Removed this code